http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeoKubernetesServiceInterp.scala

```diff
@@ -0,0 +1,412 @@
+      saveNodepool <- F.fromEither(getDefaultNodepool(clusterId, userInfo, req, ctx.now))
+      nodepool <- nodepoolQuery.saveForCluster(saveNodepool).transaction
+
+      diskOpt <- req.diskConfig.traverse(diskReq =>
+        LeoKubernetesServiceInterp.processDiskConfigRequest(
+          diskReq,
+          googleProject,
+          false,
+          userInfo,
+          authProvider,
+          serviceAccountProvider,
+          leoKubernetesConfig.diskConfig
+        )
+      )
+
+      saveApp <- F.fromEither(
+        getSaveableApp(googleProject, appName, userInfo, samResourceId, req, diskOpt, nodepool.id, ctx)
+      )
+      _ <- appQuery.save(saveApp).transaction
+////      _ <- publisherQueue.enqueue1() //TODO: queue cluster/nodepool/app creation, queue disk creation
+    } yield ()
+
+  override def getApp(
+    userInfo: UserInfo,
+    googleProject: GoogleProject,
+    appName: AppName
+  )(
+    implicit as: ApplicativeAsk[F, AppContext]
+  ): F[GetAppResponse] =
+    for {
+      ctx <- as.ask
+      appOpt <- KubernetesServiceDbQueries.getActiveFullAppByName(googleProject, appName).transaction
+      app <- F.fromOption(appOpt, AppNotFoundException(googleProject, appName, ctx.traceId))
+
+      //TODO ask SAM
+      hasPermission <- F.pure(true)
+      _ <- if (hasPermission) F.unit else F.raiseError[Unit](AppNotFoundException(googleProject, appName, ctx.traceId))
+    } yield GetAppResponse.fromDbResult(app)
+
+  override def listApp(
+    userInfo: UserInfo,
+    googleProject: Option[GoogleProject],
+    params: Map[String, String]
+  ): F[Vector[ListAppResponse]] =
+    for {
+      params <- F.fromEither(LeonardoService.processListParameters(params))
+      allClusters <- KubernetesServiceDbQueries.listFullApps(googleProject, params._1, params._2).transaction
+      //TODO: make SAM call
+      samVisibleApps <- F.pure(List[(GoogleProject, AppSamResourceId)]())
+    } yield {
+      //we construct this list of clusters by first filtering apps the user doesn't have permissions to see
+      //then we build back up by filtering nodepools without apps and clusters without nodepools
+      allClusters
+        .map { c =>
+          c.copy(nodepools =
+            c.nodepools
+              .map { n =>
+                n.copy(apps = n.apps.filter { a =>
+                  // Making the assumption that users will always be able to access apps that they create
+                  // Fix for https://github.com/DataBiosphere/leonardo/issues/821
+                  samVisibleApps
+                    .contains((c.googleProject, a.samResourceId)) || a.auditInfo.creator == userInfo.userEmail
+                })
+              }
+              .filterNot(_.apps.isEmpty)
+          )
+        }
+        .filterNot(_.nodepools.isEmpty)
+        .flatMap(c => ListAppResponse.fromCluster(c))
+        .toVector
+    }
+
+  override def deleteApp(userInfo: UserInfo, googleProject: GoogleProject, appName: AppName)(
```

Comment:
hmm... so right now I only do 'soft deletes'. I think I'll address this with my follow-up PR regarding pub/sub messaging. To ease my cognitive burden, I punted since on the difference since, right now, front leo only marks things as 'deleting', and never 'deleted'; the deleting->deleted transition will be implemented in back-leo. I think you're right, in that the follow-up ticket should add an affordance to this endpoint for it.